File: src/reqable_mcp/normalizer.py
"""Normalize Reqable report payload/HAR payload to uniform request records."""

from __future__ import annotations

import base64
import hashlib
import json
from typing import Any
from urllib.parse import parse_qs, urlparse

from .models import parse_json_if_possible, truncate_body


def _as_text(value: Any) -> str:
    if isinstance(value, str):
        return value
    if value is None:
        return ""
    if isinstance(value, (dict, list)):
        return json.dumps(value, ensure_ascii=False)
    return str(value)


def _entry_id(
    entry: dict[str, Any],
    request_body: str,
    response_body: str,
    status: int | None,
) -> str:
    req = entry.get("request", {}) or {}
    seed = "|".join(
        [
            _as_text(entry.get("_id")).strip(),
            _as_text(req.get("method", "GET")),
            _as_text(req.get("url", "")),
            _as_text(entry.get("startedDateTime", "")),
            _as_text(entry.get("time", "")),
            _as_text(status),
            hashlib.sha1(request_body.encode("utf-8", errors="ignore")).hexdigest()[:12],
            hashlib.sha1(response_body.encode("utf-8", errors="ignore")).hexdigest()[:12],
        ]
    )
    digest = hashlib.sha1(seed.encode("utf-8", errors="ignore")).hexdigest()[:24]
    return f"req-{digest}"


def _build_headers(items: Any) -> dict[str, list[str]]:
    headers: dict[str, list[str]] = {}
    if not isinstance(items, list):
        return headers
    for item in items:
        if not isinstance(item, dict):
            continue
        name = _as_text(item.get("name")).strip()
        value = _as_text(item.get("value"))
        if not name:
            continue
        headers.setdefault(name, []).append(value)
    return headers


def _first_header(headers: dict[str, list[str]], name: str) -> str | None:
    for key, values in headers.items():
        if key.lower() == name.lower():
            if values:
                return values[0]
            return None
    return None


def _header_contains(headers: dict[str, list[str]], name: str, expected: str) -> bool:
    value = _first_header(headers, name)
    return value is not None and expected.lower() in value.lower()


def _parse_query_params(parsed_url: Any) -> dict[str, str]:
    if not parsed_url.query:
        return {}
    result: dict[str, str] = {}
    for key, values in parse_qs(parsed_url.query).items():
        result[key] = values[0] if len(values) == 1 else json.dumps(values, ensure_ascii=False)
    return result


def _decode_response_text(content: dict[str, Any]) -> str:
    text = content.get("text")
    encoding = _as_text(content.get("encoding")).lower()
    text_value = _as_text(text)
    if encoding != "base64":
        return text_value
    try:
        decoded = base64.b64decode(text_value, validate=False)
        return decoded.decode("utf-8", errors="replace")
    except Exception:
        return text_value


def _websocket_candidates(entry: dict[str, Any]) -> list[dict[str, Any]]:
    for key in (
        "_webSocketMessages",
        "_websocketMessages",
        "webSocketMessages",
        "websocketMessages",
        "websocket_frames",
        "websocketFrames",
        "messages",
        "frames",
    ):
        value = entry.get(key)
        if isinstance(value, list):
            return [item for item in value if isinstance(item, dict)]
    return []


def _payload_dict(message: dict[str, Any]) -> dict[str, Any] | None:
    payload = message.get("payload")
    return payload if isinstance(payload, dict) else None


def _normalize_ws_direction(message: dict[str, Any]) -> str:
    if "fromClient" in message:
        return "outbound" if bool(message.get("fromClient")) else "inbound"
    if "outgoing" in message:
        return "outbound" if bool(message.get("outgoing")) else "inbound"
    if "flow" in message:
        try:
            flow = int(message.get("flow"))
        except (TypeError, ValueError):
            flow = None
        if flow == 0:
            return "outbound"
        if flow == 1:
            return "inbound"

    for field in ("direction", "side", "sender", "type"):
        raw = _as_text(message.get(field)).strip().lower()
        if raw in {"send", "sent", "out", "outbound", "client", "request", "up"}:
            return "outbound"
        if raw in {"receive", "received", "in", "inbound", "server", "response", "down"}:
            return "inbound"
    return "unknown"


def _normalize_ws_opcode(message: dict[str, Any]) -> int | None:
    raw = message.get("opcode")
    if raw is not None:
        try:
            return int(raw)
        except (TypeError, ValueError):
            pass

    payload = _payload_dict(message)
    payload_type = payload.get("type") if payload else None
    if payload_type == 2:
        return 1
    if payload_type == 6:
        return 8
    return None


def _normalize_ws_message_type(message: dict[str, Any], opcode: int | None) -> str | None:
    for field in ("messageType", "frameType", "frame_kind"):
        raw = _as_text(message.get(field)).strip().lower()
        if raw:
            return raw
    fallback = _as_text(message.get("type")).strip().lower()
    if fallback in {"text", "binary", "close", "ping", "pong"}:
        return fallback

    payload = _payload_dict(message)
    payload_type = payload.get("type") if payload else None
    if payload_type == 2:
        return "text"
    if payload_type == 6:
        return "close"

    if opcode == 1:
        return "text"
    if opcode == 2:
        return "binary"
    if opcode == 8:
        return "close"
    if opcode == 9:
        return "ping"
    if opcode == 10:
        return "pong"
    return None


def _extract_ws_close_details(message: dict[str, Any], data_json: Any) -> tuple[int | None, str | None]:
    candidates: list[Any] = []
    payload = _payload_dict(message)
    if isinstance(payload, dict):
        candidates.append(payload)
    candidates.append(data_json)
    candidates.append(message)

    for candidate in candidates:
        if not isinstance(candidate, dict):
            continue
        code_raw = candidate.get("code")
        code = None
        if code_raw is not None:
            try:
                code = int(code_raw)
            except (TypeError, ValueError):
                code = None
        reason_raw = candidate.get("reason")
        reason = None
        if reason_raw is not None:
            reason_text = _as_text(reason_raw).strip()
            reason = reason_text or None
        if code is not None or reason is not None:
            return code, reason
    return None, None


def _extract_ws_payload(message: dict[str, Any]) -> str:
    for key in ("data", "text", "payload", "payloadData", "body", "content"):
        if key not in message or message.get(key) is None:
            continue
        value = message.get(key)
        if isinstance(value, dict):
            for nested_key in ("text", "data", "payload", "payloadData", "body", "content"):
                if nested_key in value and value.get(nested_key) is not None:
                    return _as_text(value.get(nested_key))
        return _as_text(value)
    return ""


def normalize_websocket_message(
    message: dict[str, Any],
    max_body_size: int,
    seq: int,
) -> dict[str, Any]:
    payload_raw = _extract_ws_payload(message)
    payload, truncated = truncate_body(payload_raw, max_body_size)
    opcode = _normalize_ws_opcode(message)
    message_type = _normalize_ws_message_type(message, opcode)
    data_json = parse_json_if_possible(payload)
    close_code, close_reason = _extract_ws_close_details(message, data_json)
    is_binary = bool(message.get("binary")) or message_type == "binary" or opcode == 2
    encoding = _as_text(message.get("encoding")).strip() or None
    return {
        "seq": seq,
        "direction": _normalize_ws_direction(message),
        "timestamp": _as_text(
            message.get("time")
            or message.get("timestamp")
            or message.get("createdDateTime")
            or message.get("dateTime")
            or message.get("sentDateTime")
            or message.get("receivedDateTime")
        ).strip()
        or None,
        "opcode": opcode,
        "message_type": message_type,
        "data": payload or None,
        "data_json": data_json,
        "is_binary": is_binary,
        "encoding": encoding,
        "body_truncated": truncated,
        "close_code": close_code,
        "close_reason": close_reason,
        "raw": message,
    }


def _normalize_websocket_messages(
    entry: dict[str, Any],
    max_body_size: int,
) -> list[dict[str, Any]]:
    normalized: list[dict[str, Any]] = []
    for index, message in enumerate(_websocket_candidates(entry), start=1):
        normalized.append(normalize_websocket_message(message, max_body_size=max_body_size, seq=index))
    return normalized


def extract_entries(payload: Any) -> list[dict[str, Any]]:
    """Extract HAR entries from report payload."""
    if isinstance(payload, dict):
        if "log" in payload and isinstance(payload["log"], dict):
            entries = payload["log"].get("entries")
            if isinstance(entries, list):
                return [entry for entry in entries if isinstance(entry, dict)]
        if "entries" in payload and isinstance(payload.get("entries"), list):
            return [entry for entry in payload["entries"] if isinstance(entry, dict)]
        if isinstance(payload.get("request"), dict):
            return [payload]
        return []

    if isinstance(payload, list):
        result = []
        for item in payload:
            if isinstance(item, dict) and isinstance(item.get("request"), dict):
                result.append(item)
        return result
    return []


def normalize_entry(
    entry: dict[str, Any],
    max_body_size: int,
    source: str,
    platform: str | None = None,
    reporter_host: str | None = None,
) -> dict[str, Any]:
    req = entry.get("request", {}) or {}
    resp = entry.get("response", {}) or {}

    method = _as_text(req.get("method", "GET")).upper() or "GET"
    url = _as_text(req.get("url", "")).strip()
    parsed = urlparse(url)
    host = parsed.hostname or ""
    path = parsed.path or "/"
    query_string = parsed.query or None
    query_params = _parse_query_params(parsed)
    status: int | None
    try:
        status = int(resp.get("status")) if resp.get("status") is not None else None
    except (TypeError, ValueError):
        status = None

    request_headers = _build_headers(req.get("headers"))
    response_headers = _build_headers(resp.get("headers"))

    post_data = req.get("postData", {}) or {}
    request_body_raw = _as_text(post_data.get("text"))
    response_body_raw = _decode_response_text(resp.get("content", {}) or {})

    request_body, request_truncated = truncate_body(request_body_raw, max_body_size)
    response_body, response_truncated = truncate_body(response_body_raw, max_body_size)
    websocket_messages = _normalize_websocket_messages(entry, max_body_size)
    websocket_truncated = any(item["body_truncated"] for item in websocket_messages)
    body_truncated = request_truncated or response_truncated or websocket_truncated

    content_type = (
        _as_text(post_data.get("mimeType")).strip()
        or _first_header(request_headers, "Content-Type")
        or None
    )
    status_text = _as_text(resp.get("statusText")).strip() or None
    timestamp = _as_text(entry.get("startedDateTime")).strip() or None

    request_body_json = parse_json_if_possible(request_body)
    response_body_json = parse_json_if_possible(response_body)

    duration_ms: int | None
    try:
        duration_ms = int(float(_as_text(entry.get("time"))))
    except (TypeError, ValueError):
        duration_ms = None

    is_websocket = bool(websocket_messages)
    if not is_websocket:
        resource_type = _as_text(entry.get("_resourceType") or entry.get("resourceType")).strip().lower()
        is_websocket = resource_type == "websocket"
    if not is_websocket:
        is_websocket = parsed.scheme.lower() in {"ws", "wss"}
    if not is_websocket:
        is_websocket = (
            status == 101
            and _header_contains(request_headers, "Upgrade", "websocket")
            and _header_contains(response_headers, "Upgrade", "websocket")
        )

    has_auth = _first_header(request_headers, "Authorization") is not None
    is_https = parsed.scheme.lower() in {"https", "wss"}
    record_id = _entry_id(entry, request_body, response_body, status)

    return {
        "id": record_id,
        "method": method,
        "url": url,
        "host": host or None,
        "path": path or None,
        "query_string": query_string,
        "query_params": query_params,
        "status": status,
        "status_text": status_text,
        "duration_ms": duration_ms,
        "timestamp": timestamp,
        "request_headers": request_headers,
        "response_headers": response_headers,
        "request_body": request_body or None,
        "response_body": response_body or None,
        "request_body_json": request_body_json,
        "response_body_json": response_body_json,
        "content_type": content_type,
        "has_auth": has_auth,
        "is_https": is_https,
        "body_truncated": body_truncated,
        "remote_ip": _as_text(entry.get("serverIPAddress")).strip() or None,
        "source": source,
        "platform": platform,
        "reporter_host": reporter_host,
        "is_websocket": is_websocket,
        "websocket_messages": websocket_messages,
        "raw_entry": entry,
    }

File: go/internal/normalizer/normalizer.go
package normalizer

import (
	"crypto/sha1"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"net/url"
	"strconv"
	"strings"
)

func asText(value any) string {
	switch v := value.(type) {
	case string:
		return v
	case nil:
		return ""
	case json.Number:
		return v.String()
	case map[string]any, []any:
		data, err := json.Marshal(v)
		if err != nil {
			return fmt.Sprintf("%v", value)
		}
		return string(data)
	default:
		return fmt.Sprintf("%v", value)
	}
}

func sha1Short(text string) string {
	sum := sha1.Sum([]byte(text))
	return hex.EncodeToString(sum[:])[:12]
}

func entryID(entry map[string]any, requestBody, responseBody string, status *int) string {
	req, _ := entry["request"].(map[string]any)
	statusText := ""
	if status != nil {
		statusText = strconv.Itoa(*status)
	}
	seed := strings.Join([]string{
		strings.TrimSpace(asText(entry["_id"])),
		asText(anyMapGet(req, "method", "GET")),
		asText(anyMapGet(req, "url", "")),
		asText(entry["startedDateTime"]),
		asText(entry["time"]),
		statusText,
		sha1Short(requestBody),
		sha1Short(responseBody),
	}, "|")
	sum := sha1.Sum([]byte(seed))
	return "req-" + hex.EncodeToString(sum[:])[:24]
}

func anyMapGet(m map[string]any, key string, fallback any) any {
	if m == nil {
		return fallback
	}
	if v, ok := m[key]; ok {
		return v
	}
	return fallback
}

func buildHeaders(items any) map[string][]string {
	headers := map[string][]string{}
	list, ok := items.([]any)
	if !ok {
		return headers
	}
	for _, item := range list {
		row, ok := item.(map[string]any)
		if !ok {
			continue
		}
		name := strings.TrimSpace(asText(row["name"]))
		if name == "" {
			continue
		}
		headers[name] = append(headers[name], asText(row["value"]))
	}
	return headers
}

func firstHeader(headers map[string][]string, name string) *string {
	for key, values := range headers {
		if strings.EqualFold(key, name) {
			if len(values) == 0 {
				return nil
			}
			value := values[0]
			return &value
		}
	}
	return nil
}

func headerContains(headers map[string][]string, name, expected string) bool {
	value := firstHeader(headers, name)
	return value != nil && strings.Contains(strings.ToLower(*value), strings.ToLower(expected))
}

func parseQueryParams(parsed *url.URL) map[string]string {
	if parsed == nil || parsed.RawQuery == "" {
		return map[string]string{}
	}
	result := map[string]string{}
	values := parsed.Query()
	for key, items := range values {
		if len(items) == 1 {
			result[key] = items[0]
			continue
		}
		data, _ := json.Marshal(items)
		result[key] = string(data)
	}
	return result
}

func decodeResponseText(content map[string]any) string {
	textValue := asText(content["text"])
	encoding := strings.ToLower(asText(content["encoding"]))
	if encoding != "base64" {
		return textValue
	}
	decoded, err := base64.StdEncoding.DecodeString(textValue)
	if err != nil {
		decoded, err = base64.RawStdEncoding.DecodeString(textValue)
		if err != nil {
			return textValue
		}
	}
	return string(decoded)
}

func websocketCandidates(entry map[string]any) []map[string]any {
	for _, key := range []string{"_webSocketMessages", "_websocketMessages", "webSocketMessages", "websocketMessages", "websocket_frames", "websocketFrames", "messages", "frames"} {
		if list, ok := entry[key].([]any); ok {
			result := make([]map[string]any, 0, len(list))
			for _, item := range list {
				if msg, ok := item.(map[string]any); ok {
					result = append(result, msg)
				}
			}
			return result
		}
	}
	return nil
}

func payloadDict(message map[string]any) map[string]any {
	payload, _ := message["payload"].(map[string]any)
	return payload
}

func normalizeWSDirection(message map[string]any) string {
	if v, ok := message["fromClient"]; ok {
		if toBool(v) {
			return "outbound"
		}
		return "inbound"
	}
	if v, ok := message["outgoing"]; ok {
		if toBool(v) {
			return "outbound"
		}
		return "inbound"
	}
	if v, ok := message["flow"]; ok {
		if i, ok := toInt(v); ok {
			if i == 0 {
				return "outbound"
			}
			if i == 1 {
				return "inbound"
			}
		}
	}
	for _, field := range []string{"direction", "side", "sender", "type"} {
		raw := strings.ToLower(strings.TrimSpace(asText(message[field])))
		switch raw {
		case "send", "sent", "out", "outbound", "client", "request", "up":
			return "outbound"
		case "receive", "received", "in", "inbound", "server", "response", "down":
			return "inbound"
		}
	}
	return "unknown"
}

func normalizeWSOpcode(message map[string]any) *int {
	if raw, ok := message["opcode"]; ok {
		if i, ok := toInt(raw); ok {
			return &i
		}
	}
	payload := payloadDict(message)
	if payload != nil {
		if t, ok := toInt(payload["type"]); ok {
			if t == 2 {
				value := 1
				return &value
			}
			if t == 6 {
				value := 8
				return &value
			}
		}
	}
	return nil
}

func normalizeWSMessageType(message map[string]any, opcode *int) *string {
	for _, field := range []string{"messageType", "frameType", "frame_kind"} {
		raw := strings.ToLower(strings.TrimSpace(asText(message[field])))
		if raw != "" {
			return &raw
		}
	}
	fallback := strings.ToLower(strings.TrimSpace(asText(message["type"])))
	switch fallback {
	case "text", "binary", "close", "ping", "pong":
		return &fallback
	}
	payload := payloadDict(message)
	if payload != nil {
		if t, ok := toInt(payload["type"]); ok {
			switch t {
			case 2:
				raw := "text"
				return &raw
			case 6:
				raw := "close"
				return &raw
			}
		}
	}
	if opcode != nil {
		switch *opcode {
		case 1:
			raw := "text"
			return &raw
		case 2:
			raw := "binary"
			return &raw
		case 8:
			raw := "close"
			return &raw
		case 9:
			raw := "ping"
			return &raw
		case 10:
			raw := "pong"
			return &raw
		}
	}
	return nil
}

func extractWSCloseDetails(message map[string]any, dataJSON any) (*int, *string) {
	candidates := []any{payloadDict(message), dataJSON, message}
	for _, candidate := range candidates {
		item, ok := candidate.(map[string]any)
		if !ok {
			continue
		}
		var codePtr *int
		if code, ok := toInt(item["code"]); ok {
			codePtr = &code
		}
		var reasonPtr *string
		if value, exists := item["reason"]; exists && value != nil {
			reason := strings.TrimSpace(asText(value))
			if reason != "" {
				reasonPtr = &reason
			}
		}
		if codePtr != nil || reasonPtr != nil {
			return codePtr, reasonPtr
		}
	}
	return nil, nil
}

func extractWSPayload(message map[string]any) string {
	for _, key := range []string{"data", "text", "payload", "payloadData", "body", "content"} {
		value, ok := message[key]
		if !ok || value == nil {
			continue
		}
		if nested, ok := value.(map[string]any); ok {
			for _, nestedKey := range []string{"text", "data", "payload", "payloadData", "body", "content"} {
				if nestedValue, ok := nested[nestedKey]; ok && nestedValue != nil {
					return asText(nestedValue)
				}
			}
		}
		return asText(value)
	}
	return ""
}

func ParseJSONIfPossible(text string) any {
	if text == "" {
		return nil
	}
	var result any
	if err := json.Unmarshal([]byte(text), &result); err != nil {
		return nil
	}
	return result
}

func TruncateBody(body string, maxLength int) (string, bool) {
	if len(body) <= maxLength {
		return body, false
	}
	return body[:maxLength] + "...[truncated]", true
}

func toBool(value any) bool {
	switch v := value.(type) {
	case bool:
		return v
	case string:
		parsed, _ := strconv.ParseBool(v)
		return parsed
	case float64:
		return v != 0
	case int:
		return v != 0
	case int64:
		return v != 0
	default:
		return false
	}
}

func toInt(value any) (int, bool) {
	switch v := value.(type) {
	case int:
		return v, true
	case int8:
		return int(v), true
	case int16:
		return int(v), true
	case int32:
		return int(v), true
	case int64:
		return int(v), true
	case uint:
		return int(v), true
	case uint32:
		return int(v), true
	case uint64:
		return int(v), true
	case float64:
		return int(v), true
	case float32:
		return int(v), true
	case json.Number:
		i, err := v.Int64()
		if err == nil {
			return int(i), true
		}
		f, err := v.Float64()
		if err == nil {
			return int(f), true
		}
		return 0, false
	case string:
		if strings.TrimSpace(v) == "" {
			return 0, false
		}
		i, err := strconv.Atoi(strings.TrimSpace(v))
		return i, err == nil
	default:
		return 0, false
	}
}

func NormalizeWebSocketMessage(message map[string]any, maxBodySize int, seq int) map[string]any {
	payloadRaw := extractWSPayload(message)
	payload, truncated := TruncateBody(payloadRaw, maxBodySize)
	opcode := normalizeWSOpcode(message)
	messageType := normalizeWSMessageType(message, opcode)
	dataJSON := ParseJSONIfPossible(payload)
	closeCode, closeReason := extractWSCloseDetails(message, dataJSON)
	isBinary := toBool(message["binary"])
	if !isBinary && messageType != nil && *messageType == "binary" {
		isBinary = true
	}
	if !isBinary && opcode != nil && *opcode == 2 {
		isBinary = true
	}
	encoding := strings.TrimSpace(asText(message["encoding"]))
	var encodingPtr *string
	if encoding != "" {
		encodingPtr = &encoding
	}
	timestamp := strings.TrimSpace(asText(firstNonNil(
		message["time"],
		message["timestamp"],
		message["createdDateTime"],
		message["dateTime"],
		message["sentDateTime"],
		message["receivedDateTime"],
	)))
	var timestampPtr *string
	if timestamp != "" {
		timestampPtr = &timestamp
	}
	return map[string]any{
		"seq":            seq,
		"direction":      normalizeWSDirection(message),
		"timestamp":      timestampPtr,
		"opcode":         opcode,
		"message_type":   messageType,
		"data":           nilIfEmpty(payload),
		"data_json":      dataJSON,
		"is_binary":      isBinary,
		"encoding":       encodingPtr,
		"body_truncated": truncated,
		"close_code":     closeCode,
		"close_reason":   closeReason,
		"raw":            message,
	}
}

func firstNonNil(values ...any) any {
	for _, value := range values {
		if value != nil && asText(value) != "" {
			return value
		}
	}
	return nil
}

func nilIfEmpty(text string) any {
	if text == "" {
		return nil
	}
	return text
}

func normalizeWebSocketMessages(entry map[string]any, maxBodySize int) []map[string]any {
	candidates := websocketCandidates(entry)
	result := make([]map[string]any, 0, len(candidates))
	for idx, message := range candidates {
		result = append(result, NormalizeWebSocketMessage(message, maxBodySize, idx+1))
	}
	return result
}

func ExtractEntries(payload any) []map[string]any {
	switch v := payload.(type) {
	case map[string]any:
		if logValue, ok := v["log"].(map[string]any); ok {
			if entries, ok := logValue["entries"].([]any); ok {
				return filterRequestEntries(entries)
			}
		}
		if entries, ok := v["entries"].([]any); ok {
			return filterRequestEntries(entries)
		}
		if _, ok := v["request"].(map[string]any); ok {
			return []map[string]any{v}
		}
	case []any:
		return filterRequestEntries(v)
	}
	return nil
}

func filterRequestEntries(items []any) []map[string]any {
	result := make([]map[string]any, 0, len(items))
	for _, item := range items {
		entry, ok := item.(map[string]any)
		if !ok {
			continue
		}
		if _, hasReq := entry["request"].(map[string]any); hasReq {
			result = append(result, entry)
		}
	}
	return result
}

func NormalizeEntry(entry map[string]any, maxBodySize int, source string, platform, reporterHost *string) map[string]any {
	req, _ := entry["request"].(map[string]any)
	resp, _ := entry["response"].(map[string]any)
	method := strings.ToUpper(asText(anyMapGet(req, "method", "GET")))
	if method == "" {
		method = "GET"
	}
	rawURL := strings.TrimSpace(asText(anyMapGet(req, "url", "")))
	parsed, _ := url.Parse(rawURL)
	host := ""
	path := "/"
	queryString := any(nil)
	queryParams := map[string]string{}
	isHTTPS := false
	if parsed != nil {
		host = parsed.Hostname()
		if parsed.Path != "" {
			path = parsed.Path
		}
		if parsed.RawQuery != "" {
			queryString = parsed.RawQuery
		}
		queryParams = parseQueryParams(parsed)
		isHTTPS = strings.EqualFold(parsed.Scheme, "https") || strings.EqualFold(parsed.Scheme, "wss")
	}
	var status *int
	if resp != nil {
		if i, ok := toInt(resp["status"]); ok {
			status = &i
		}
	}
	requestHeaders := buildHeaders(anyMapGet(req, "headers", nil))
	responseHeaders := buildHeaders(anyMapGet(resp, "headers", nil))
	postData, _ := anyMapGet(req, "postData", map[string]any{}).(map[string]any)
	requestBodyRaw := asText(anyMapGet(postData, "text", ""))
	responseContent, _ := anyMapGet(resp, "content", map[string]any{}).(map[string]any)
	responseBodyRaw := decodeResponseText(responseContent)
	requestBody, requestTruncated := TruncateBody(requestBodyRaw, maxBodySize)
	responseBody, responseTruncated := TruncateBody(responseBodyRaw, maxBodySize)
	websocketMessages := normalizeWebSocketMessages(entry, maxBodySize)
	websocketTruncated := false
	for _, item := range websocketMessages {
		if b, _ := item["body_truncated"].(bool); b {
			websocketTruncated = true
			break
		}
	}
	bodyTruncated := requestTruncated || responseTruncated || websocketTruncated
	contentType := strings.TrimSpace(asText(anyMapGet(postData, "mimeType", "")))
	if contentType == "" {
		if header := firstHeader(requestHeaders, "Content-Type"); header != nil {
			contentType = *header
		}
	}
	var contentTypePtr *string
	if contentType != "" {
		contentTypePtr = &contentType
	}
	statusText := strings.TrimSpace(asText(anyMapGet(resp, "statusText", "")))
	var statusTextPtr *string
	if statusText != "" {
		statusTextPtr = &statusText
	}
	timestamp := strings.TrimSpace(asText(entry["startedDateTime"]))
	var timestampPtr *string
	if timestamp != "" {
		timestampPtr = &timestamp
	}
	requestBodyJSON := ParseJSONIfPossible(requestBody)
	responseBodyJSON := ParseJSONIfPossible(responseBody)
	var durationMS *int
	if raw := strings.TrimSpace(asText(entry["time"])); raw != "" {
		if f, err := strconv.ParseFloat(raw, 64); err == nil {
			value := int(f)
			durationMS = &value
		}
	}
	isWebSocket := len(websocketMessages) > 0
	if !isWebSocket {
		resourceType := strings.ToLower(strings.TrimSpace(asText(firstNonNil(entry["_resourceType"], entry["resourceType"]))))
		isWebSocket = resourceType == "websocket"
	}
	if !isWebSocket && parsed != nil {
		isWebSocket = strings.EqualFold(parsed.Scheme, "ws") || strings.EqualFold(parsed.Scheme, "wss")
	}
	if !isWebSocket && status != nil {
		isWebSocket = *status == 101 && headerContains(requestHeaders, "Upgrade", "websocket") && headerContains(responseHeaders, "Upgrade", "websocket")
	}
	hasAuth := firstHeader(requestHeaders, "Authorization") != nil
	var hostPtr, pathPtr, remoteIPPtr, requestBodyPtr, responseBodyPtr *string
	if host != "" {
		hostPtr = &host
	}
	if path != "" {
		pathPtr = &path
	}
	remoteIP := strings.TrimSpace(asText(entry["serverIPAddress"]))
	if remoteIP != "" {
		remoteIPPtr = &remoteIP
	}
	if requestBody != "" {
		requestBodyPtr = &requestBody
	}
	if responseBody != "" {
		responseBodyPtr = &responseBody
	}
	return map[string]any{
		"id":                 entryID(entry, requestBody, responseBody, status),
		"method":             method,
		"url":                rawURL,
		"host":               hostPtr,
		"path":               pathPtr,
		"query_string":       queryString,
		"query_params":       queryParams,
		"status":             status,
		"status_text":        statusTextPtr,
		"duration_ms":        durationMS,
		"timestamp":          timestampPtr,
		"request_headers":    requestHeaders,
		"response_headers":   responseHeaders,
		"request_body":       requestBodyPtr,
		"response_body":      responseBodyPtr,
		"request_body_json":  requestBodyJSON,
		"response_body_json": responseBodyJSON,
		"content_type":       contentTypePtr,
		"has_auth":           hasAuth,
		"is_https":           isHTTPS,
		"body_truncated":     bodyTruncated,
		"remote_ip":          remoteIPPtr,
		"source":             source,
		"platform":           platform,
		"reporter_host":      reporterHost,
		"is_websocket":       isWebSocket,
		"websocket_messages": websocketMessages,
		"raw_entry":          entry,
	}
}
